File: Card.py
############################################# Card Model #############################################

SUITS = ('hearts', 'diamonds', 'clubs', 'spades')

# Ranks in ascending trick taking strength: a rank's index here is its Card.order.
# NOTE: deliberately different from the point values below (a 7 beats a King but an Ace beats both).
RANKS = ('2', '3', '4', '5', '6', 'Q', 'J', 'K', '7', 'A')

# Points a card is worth to whoever captures it. Ranks left out are worth nothing.
RANK_VALUES = {'A': 11, '7': 10, 'K': 4, 'J': 3, 'Q': 2}

# Lookup tables derived from the above, so no module has to hardcode them again.
SUIT_INDEX = {suit: index for index, suit in enumerate(SUITS)}
RANK_ORDER = {rank: index for index, rank in enumerate(RANKS)}
ORDER_VALUES = tuple(RANK_VALUES.get(rank, 0) for rank in RANKS)   # indexed by Card.order


class Card:
    '''
        Card ->
            - name: card name (e.g. "A_of_hearts")
            - suit: card suit (hearts, diamonds, clubs, spades)
            - rank: card rank (2, 3, 4, 5, 6, Q, J, K, 7, A)
            - order: trick taking strength (0 - 9), the rank's index in RANKS
            - value: points the card is worth (0, 2, 3, 4, 10, 11)
    '''

    __slots__ = ('name', 'suit', 'rank', 'order', 'value')

    def __init__(self, suit: str, rank: str) -> None:
        self.suit = suit
        self.rank = rank
        self.name = f'{rank}_of_{suit}'
        self.order = RANK_ORDER[rank]
        self.value = RANK_VALUES.get(rank, 0)

    def beats(self, current_best: 'Card', trump_suit: str) -> bool:
        '''
            Whether this card would take a round currently being won by current_best.

            Only a higher card of the same suit or a trump played on a non trump wins,
            so a card of the leading suit never beats a round that has been cut.
        '''

        if self.suit == current_best.suit:
            return self.order > current_best.order

        return self.suit == trump_suit

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Card):
            return NotImplemented

        return self.suit == other.suit and self.rank == other.rank

    def __hash__(self) -> int:
        return hash((self.suit, self.rank))

    def __str__(self) -> str:
        return self.name

    def __repr__(self) -> str:
        return f'Card({self.name})'


def new_deck() -> list[Card]:
    '''
        Build the 40 card Sueca deck: every rank of every suit
    '''

    return [Card(suit, rank) for rank in RANKS for suit in SUITS]

File: android/engine/src/main/kotlin/pt/up/fe/asma/sueca/engine/Card.kt
package pt.up.fe.asma.sueca.engine

/**
 * Card model, ported from `Card.py`.
 *
 * Two number systems live on a card and are easy to confuse:
 *  - [Rank.points] is what the card *scores* for whoever captures it (A=11, 7=10, K=4, J=3, Q=2).
 *  - [Rank.order] is its *trick strength* (0..9) over the sequence 2 3 4 5 6 Q J K 7 A.
 *
 * Round winners are decided by order, scores by points. This file is the single source of
 * truth for both, exactly like `Card.py` is on the Python side.
 */
enum class Suit(val id: String, val symbol: String, val isRed: Boolean) {
    HEARTS("hearts", "♥", true),
    DIAMONDS("diamonds", "♦", true),
    CLUBS("clubs", "♣", false),
    SPADES("spades", "♠", false);

    companion object {
        fun fromId(id: String): Suit? = entries.firstOrNull { it.id.equals(id, ignoreCase = true) }
    }
}

/** Ranks in ascending trick taking strength: the ordinal *is* [order]. */
enum class Rank(val label: String, val points: Int) {
    TWO("2", 0),
    THREE("3", 0),
    FOUR("4", 0),
    FIVE("5", 0),
    SIX("6", 0),
    QUEEN("Q", 2),
    JACK("J", 3),
    KING("K", 4),
    SEVEN("7", 10),
    ACE("A", 11);

    /** Trick taking strength, 0 (weakest) to 9 (strongest). */
    val order: Int get() = ordinal

    companion object {
        fun fromLabel(label: String): Rank? = entries.firstOrNull { it.label.equals(label, ignoreCase = true) }
    }
}

data class Card(val suit: Suit, val rank: Rank) {

    /** Trick taking strength (0..9). Never compare this directly to decide a round, use [beats]. */
    val order: Int get() = rank.order

    /** Points this card is worth to whoever captures it. */
    val points: Int get() = rank.points

    /** `A_of_hearts`, the same identifier the Python simulator writes into its logs. */
    val id: String get() = "${rank.label}_of_${suit.id}"

    /** Compact label for the UI, e.g. `Q♠`. */
    val label: String get() = "${rank.label}${suit.symbol}"

    /** Dense index 0..39, used to address the belief tables. */
    val index: Int get() = suit.ordinal * RANK_COUNT + rank.order

    /**
     * Whether this card would take a round currently being won by [currentBest].
     *
     * Only a higher card of the same suit, or a trump played on a non trump, wins. A card of
     * the leading suit therefore never beats a round that has already been cut, which is why
     * comparing [order] on its own is always wrong.
     */
    fun beats(currentBest: Card, trump: Suit): Boolean =
        if (suit == currentBest.suit) order > currentBest.order else suit == trump

    override fun toString(): String = id

    companion object {
        fun of(suit: Suit, rank: Rank) = Card(suit, rank)

        /** Parses `A_of_hearts` or `AH` / `7♦` style labels. Returns null when unparseable. */
        fun parse(text: String): Card? {
            val trimmed = text.trim()
            if (trimmed.contains("_of_")) {
                val (rankPart, suitPart) = trimmed.split("_of_", limit = 2)
                val rank = Rank.fromLabel(rankPart) ?: return null
                val suit = Suit.fromId(suitPart) ?: return null
                return Card(suit, rank)
            }
            if (trimmed.length < 2) return null
            val rank = Rank.fromLabel(trimmed.dropLast(1)) ?: return null
            val suitChar = trimmed.last()
            val suit = Suit.entries.firstOrNull {
                it.symbol[0] == suitChar || it.id[0].equals(suitChar, ignoreCase = true)
            } ?: return null
            return Card(suit, rank)
        }

        fun fromIndex(index: Int): Card = Card(Suit.entries[index / RANK_COUNT], Rank.entries[index % RANK_COUNT])
    }
}

const val SUIT_COUNT = 4
const val RANK_COUNT = 10
const val CARDS_PER_HAND = 10
const val ROUNDS_PER_GAME = 10
const val PLAYER_COUNT = 4

/** Every point in the deck: 4 x (11 + 10 + 4 + 3 + 2) = 120. */
const val TOTAL_POINTS = 120

/** Points per [Card.order], the equivalent of `ORDER_VALUES` in `Card.py`. */
val ORDER_POINTS: IntArray = IntArray(RANK_COUNT) { Rank.entries[it].points }

/**
 * The 40 card Sueca deck, in the same order `Card.new_deck()` builds it (rank major), so a
 * seeded deal here matches a seeded deal there.
 */
val FULL_DECK: List<Card> = buildList {
    for (rank in Rank.entries) {
        for (suit in Suit.entries) {
            add(Card(suit, rank))
        }
    }
}

fun newDeck(): MutableList<Card> = FULL_DECK.toMutableList()
